File: spherepop-go/wire.go
package spherepop

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"sort"
	"unicode/utf8"
)

var wireMagic = [8]byte{'S', 'P', 'H', 'I', 'S', 'T', '1', 0}

type WireWorld struct {
	InitialOptionSpace []ObjectID
	CertifiedRules     []RuleID
	History            History
}

func putString(out *bytes.Buffer, value string) error {
	data := []byte(value)
	if uint64(len(data)) > uint64(^uint32(0)) {
		return errors.New("SPHIST/1 string too long")
	}
	_ = binary.Write(out, binary.BigEndian, uint32(len(data)))
	out.Write(data)
	return nil
}

func EncodeHistory(initial []ObjectID, rules []RuleID, history History) ([]byte, error) {
	omega := append([]ObjectID(nil), initial...)
	sort.Slice(omega, func(i, j int) bool { return omega[i] < omega[j] })
	for i := 1; i < len(omega); i++ {
		if omega[i] == omega[i-1] { return nil, errors.New("duplicate initial option") }
	}
	certified := append([]RuleID(nil), rules...)
	sort.Slice(certified, func(i, j int) bool { return certified[i] < certified[j] })
	for i := 1; i < len(certified); i++ {
		if certified[i] == certified[i-1] { return nil, errors.New("duplicate certified rule") }
	}
	if uint64(len(omega)) > uint64(^uint32(0)) || uint64(len(certified)) > uint64(^uint32(0)) || uint64(history.Len()) > uint64(^uint32(0)) {
		return nil, errors.New("SPHIST/1 collection too large")
	}
	out := bytes.NewBuffer(nil)
	out.Write(wireMagic[:])
	_ = binary.Write(out, binary.BigEndian, uint32(len(omega)))
	for _, id := range omega { _ = binary.Write(out, binary.BigEndian, uint64(id)) }
	_ = binary.Write(out, binary.BigEndian, uint32(len(certified)))
	for _, rule := range certified { if err := putString(out, string(rule)); err != nil { return nil, err } }
	_ = binary.Write(out, binary.BigEndian, uint32(history.Len()))
	for _, event := range history.events {
		out.WriteByte(byte(event.Kind))
		switch event.Kind {
		case Pop:
			_ = binary.Write(out, binary.BigEndian, uint64(*event.A))
		case Refuse:
			_ = binary.Write(out, binary.BigEndian, uint64(*event.A))
			if event.B == nil { out.WriteByte(0) } else { out.WriteByte(1); _ = binary.Write(out, binary.BigEndian, uint64(*event.B)) }
			if err := putString(out, *event.Reason); err != nil { return nil, err }
		case Bind:
			_ = binary.Write(out, binary.BigEndian, uint64(*event.A))
			_ = binary.Write(out, binary.BigEndian, uint64(*event.B))
			if err := putString(out, *event.Tag); err != nil { return nil, err }
		case Collapse:
			if err := putString(out, string(*event.Rule)); err != nil { return nil, err }
		default:
			return nil, fmt.Errorf("unknown event kind %d", event.Kind)
		}
	}
	return out.Bytes(), nil
}

type wireReader struct { data []byte; offset int }
func (r *wireReader) take(n int) ([]byte, error) { if n < 0 || r.offset+n > len(r.data) { return nil, errors.New("truncated SPHIST/1 envelope") }; b:=r.data[r.offset:r.offset+n]; r.offset+=n; return b,nil }
func (r *wireReader) u8() (byte,error) { b,e:=r.take(1); if e!=nil{return 0,e}; return b[0],nil }
func (r *wireReader) u32() (uint32,error) { b,e:=r.take(4); if e!=nil{return 0,e}; return binary.BigEndian.Uint32(b),nil }
func (r *wireReader) u64() (uint64,error) { b,e:=r.take(8); if e!=nil{return 0,e}; return binary.BigEndian.Uint64(b),nil }
func (r *wireReader) str() (string,error) { n,e:=r.u32(); if e!=nil{return "",e}; b,e:=r.take(int(n)); if e!=nil{return "",e}; if !utf8.Valid(b){return "",errors.New("invalid UTF-8 in SPHIST/1 envelope")}; return string(b),nil }

func DecodeHistory(data []byte) (WireWorld, error) {
	r := &wireReader{data:data}
	magic, err := r.take(len(wireMagic)); if err != nil || !bytes.Equal(magic, wireMagic[:]) { return WireWorld{}, errors.New("invalid SPHIST/1 magic") }
	nOmega,err:=r.u32(); if err!=nil{return WireWorld{},err}; omega:=make([]ObjectID,nOmega)
	for i:=range omega { v,e:=r.u64(); if e!=nil{return WireWorld{},e}; omega[i]=ObjectID(v); if i>0 && omega[i]<=omega[i-1]{return WireWorld{},errors.New("non-canonical option ordering")} }
	nRules,err:=r.u32(); if err!=nil{return WireWorld{},err}; rules:=make([]RuleID,nRules)
	for i:=range rules { v,e:=r.str(); if e!=nil{return WireWorld{},e}; rules[i]=RuleID(v); if i>0 && rules[i]<=rules[i-1]{return WireWorld{},errors.New("non-canonical rule ordering")} }
	nEvents,err:=r.u32(); if err!=nil{return WireWorld{},err}; history:=NewHistory()
	for i:=uint32(0); i<nEvents; i++ {
		kind,e:=r.u8(); if e!=nil{return WireWorld{},e}; var event Event
		switch EventKind(kind) {
		case Pop: a,e:=r.u64(); if e!=nil{return WireWorld{},e}; event=PopEvent(ObjectID(a))
		case Refuse:
			a,e:=r.u64(); if e!=nil{return WireWorld{},e}; flag,e:=r.u8(); if e!=nil{return WireWorld{},e}; if flag>1{return WireWorld{},errors.New("invalid Refuse has_b flag")}; var b uint64; if flag==1 { b,e=r.u64(); if e!=nil{return WireWorld{},e} }; reason,e:=r.str(); if e!=nil{return WireWorld{},e}; if flag==1 { event=RefuseBindEvent(ObjectID(a),ObjectID(b),reason) } else { event=RefuseEvent(ObjectID(a),reason) }
		case Bind: a,e:=r.u64(); if e!=nil{return WireWorld{},e}; b,e:=r.u64(); if e!=nil{return WireWorld{},e}; tag,e:=r.str(); if e!=nil{return WireWorld{},e}; event=BindEvent(ObjectID(a),ObjectID(b),tag)
		case Collapse: rule,e:=r.str(); if e!=nil{return WireWorld{},e}; event=CollapseEvent(RuleID(rule))
		default: return WireWorld{},fmt.Errorf("invalid event kind %d",kind)
		}
		event.Pos=LogPos(i); history.append(event)
	}
	if r.offset!=len(data){return WireWorld{},errors.New("trailing bytes in SPHIST/1 envelope")}
	return WireWorld{omega,rules,history},nil
}

func FNV1a64(data []byte) string {
	var digest uint64 = 0xcbf29ce484222325
	for _, b := range data { digest ^= uint64(b); digest *= 0x100000001b3 }
	return fmt.Sprintf("%016x",digest)
}

File: spherepop-kernel/src/wire.rs
//! Canonical SPHIST/1 history envelope (see experiments/flat/HISTORY-WIRE-V1.md).

use crate::{Event, EventKind, History, ObjectId, RuleId};
use std::collections::HashSet;

const MAGIC: &[u8; 8] = b"SPHIST1\0";

fn put_u32(out: &mut Vec<u8>, value: usize) -> Result<(), String> {
    let value = u32::try_from(value).map_err(|_| "SPHIST/1 collection too large".to_string())?;
    out.extend_from_slice(&value.to_be_bytes());
    Ok(())
}

fn put_string(out: &mut Vec<u8>, value: &str) -> Result<(), String> {
    put_u32(out, value.as_bytes().len())?;
    out.extend_from_slice(value.as_bytes());
    Ok(())
}

pub fn encode_history(
    initial: impl IntoIterator<Item = ObjectId>,
    rules: impl IntoIterator<Item = RuleId>,
    history: &History,
) -> Result<Vec<u8>, String> {
    let mut omega: Vec<_> = initial.into_iter().collect();
    omega.sort_unstable();
    if omega.windows(2).any(|pair| pair[0] == pair[1]) { return Err("duplicate initial option".into()); }
    let mut certified: Vec<_> = rules.into_iter().collect();
    certified.sort_unstable();
    if certified.windows(2).any(|pair| pair[0] == pair[1]) { return Err("duplicate certified rule".into()); }

    let mut out = MAGIC.to_vec();
    put_u32(&mut out, omega.len())?;
    for id in omega { out.extend_from_slice(&id.to_be_bytes()); }
    put_u32(&mut out, certified.len())?;
    for rule in certified { put_string(&mut out, rule)?; }
    put_u32(&mut out, history.len())?;
    for event in history.as_slice() {
        out.push(event.kind as u8);
        match event.kind {
            EventKind::Pop => out.extend_from_slice(&event.a.ok_or("Pop missing a")?.to_be_bytes()),
            EventKind::Refuse => {
                out.extend_from_slice(&event.a.ok_or("Refuse missing a")?.to_be_bytes());
                match event.b { Some(b) => { out.push(1); out.extend_from_slice(&b.to_be_bytes()); }, None => out.push(0) }
                put_string(&mut out, event.reason.as_deref().ok_or("Refuse missing reason")?)?;
            }
            EventKind::Bind => {
                out.extend_from_slice(&event.a.ok_or("Bind missing a")?.to_be_bytes());
                out.extend_from_slice(&event.b.ok_or("Bind missing b")?.to_be_bytes());
                put_string(&mut out, event.tag.as_deref().ok_or("Bind missing tag")?)?;
            }
            EventKind::Collapse => put_string(&mut out, event.rule.ok_or("Collapse missing rule")?)?,
        }
    }
    Ok(out)
}

struct Reader<'a> { data: &'a [u8], offset: usize }
impl<'a> Reader<'a> {
    fn take(&mut self, n: usize) -> Result<&'a [u8], String> { let end=self.offset.checked_add(n).ok_or("truncated SPHIST/1 envelope")?; if end>self.data.len(){return Err("truncated SPHIST/1 envelope".into())}; let value=&self.data[self.offset..end]; self.offset=end; Ok(value) }
    fn u8(&mut self) -> Result<u8,String> { Ok(self.take(1)?[0]) }
    fn u32(&mut self) -> Result<u32,String> { Ok(u32::from_be_bytes(self.take(4)?.try_into().unwrap())) }
    fn u64(&mut self) -> Result<u64,String> { Ok(u64::from_be_bytes(self.take(8)?.try_into().unwrap())) }
    fn string(&mut self) -> Result<&'static str,String> { let n=self.u32()? as usize; let value=std::str::from_utf8(self.take(n)?).map_err(|_|"invalid UTF-8 in SPHIST/1 envelope")?; Ok(Box::leak(value.to_string().into_boxed_str())) }
}

pub struct WireWorld { pub initial_option_space: HashSet<ObjectId>, pub certified_rules: Vec<RuleId>, pub history: History }

pub fn decode_history(data: &[u8]) -> Result<WireWorld, String> {
    let mut r=Reader{data,offset:0};
    if r.take(MAGIC.len())? != MAGIC { return Err("invalid SPHIST/1 magic".into()); }
    let mut omega=Vec::new(); for _ in 0..r.u32()? { omega.push(r.u64()?); }
    if omega.windows(2).any(|p|p[0]>=p[1]) { return Err("non-canonical option ordering".into()); }
    let mut rules=Vec::new(); for _ in 0..r.u32()? { rules.push(r.string()?); }
    if rules.windows(2).any(|p|p[0]>=p[1]) { return Err("non-canonical rule ordering".into()); }
    let mut history=History::new();
    for pos in 0..r.u32()? {
        let mut event=match r.u8()? {
            0 => Event::pop(r.u64()?),
            1 => { let a=r.u64()?; let flag=r.u8()?; let b=match flag {0=>None,1=>Some(r.u64()?),_=>return Err("invalid Refuse has_b flag".into())}; let reason=r.string()?.to_string(); match b {Some(b)=>Event::refuse_bind(a,b,reason),None=>Event::refuse(a,reason)} },
            2 => { let a=r.u64()?; let b=r.u64()?; Event::bind(a,b,r.string()?.to_string()) },
            3 => Event::collapse(r.string()?),
            kind => return Err(format!("invalid event kind {kind}")),
        };
        event.pos=pos as u64;
        history.push(event);
    }
    if r.offset != data.len() { return Err("trailing bytes in SPHIST/1 envelope".into()); }
    Ok(WireWorld{initial_option_space:omega.into_iter().collect(),certified_rules:rules,history})
}

pub fn fnv1a64(data: &[u8]) -> String {
    let mut digest=0xcbf29ce484222325u64;
    for byte in data { digest ^= *byte as u64; digest= digest.wrapping_mul(0x100000001b3); }
    format!("{digest:016x}")
}
